docs/dev/HOCInterpreter/HOCInterpreter.md

```diff
@@ -0,0 +1,496 @@
+
+  
+
+So given the following HOC call: 
+
+    
+    oc>printf("%s equals %d", "one plus two", 1+2)
+    
+
+the parsing will parse the arguments and the interpreter will perform **_call()_.** At this point, program counter _**pc**_  points to the **_printf_** symbol table entry and just after we have the number of arguments _**nargs.**_ We now push and setup a new _**Frame**_:
+
+  ```cpp  
+    void call(void)    /* call a function */
+    {
+       int i, isec;
+       Symbol *sp = pc[0].sym;    /* symbol table entry for printf*/
+       
+       /* stack a new frame */           
+       if (++fp >= framelast) {
+          --fp;
+          execerror(sp->name, "call nested too deeply, increase with -NFRAME framesize option");
+       }
+       fp->sp = sp;
+       fp->nargs = pc[1].i;
+       fp->retpc = pc + 2; /* This is where we return to after the printf call */
+    ...
+       fp->argn = stackp - 2;  /* pointer to the last argument */
+  ```  
+
+By inspecting the frame we notice:
+
+  ```cpp    
+    (lldb) p *fp
+    (Frame) $2 = {
+      sp = 0x0000000100d07dc0
+      retpc = 0x0000000100b6c060
+      argn = 0x0000000101008c20
+      nargs = 3
+      iter_stmt_begin = 0x0000000000000000
+      iter_stmt_ob = 0x0000000000000000
+      ob = 0x0000000000000000
+    }
+    
+    (lldb) p *fp->sp
+    (Symbol) $3 = {
+      name = 0x0000000100d07e00 "printf"
+      type = 280 // this is the type of the symbol, which corresponds to a builtin function: FUN_BLTIN = 280
+      subtype = 0
+      public = 0
+      defined_on_the_fly = 0
+      u = {
+        oboff = 13663760
+        pval = 0x0000000100d07e10
+        object_ = 0x0000000100d07e10
+        cstr = 0x0000000100d07e10 "??'"
+        pnum = 0x0000000100d07e10
+        pvalint = 0x0000000100d07e10
+        pvalfloat = 0x0000000100d07e10
+        u_auto = 13663760
+        ptr = 0x0000000100d07e10 (0x0000000100d07e10)
+        u_proc = 0x0000000100d07e10
+        rng = (type = 32272, index = 1)
+        ppsym = 0x0000000100d07e10
+        template = 0x0000000100d07e10
+        sym = 0x0000000100d07e10
+      }
+      s_varn = 0
+      arayinfo = 0x0000000000000000
+      extra = 0x0000000000000000
+      next = 0x0000000100d07e30
+    }
+    
+    (lldb) p *fp->argn
+    (Datum) $4 = {
+      val = 3 // value of the last argument (1+2) has already been computed
+      sym = 0x4008000000000000
+      i = 0
+      pval = 0x4008000000000000
+      pobj = 0x4008000000000000
+      obj = 0x4008000000000000
+      pstr = 0x4008000000000000
+      itm = 0x4008000000000000
+      lst = 0x4008000000000000
+      _pvoid = 0x4008000000000000
+    }
+  ```  
+
+_**call()**_ will continue on to check the symbol, which  tells us we are calling a builtin function (_**FUN\_BLTIN = 280**_, see _**fp**_ above), and call its function pointer mapped to **_hoc\_PRintf_**: 
```

Comment:
So I think you're right here, but I don't quite understand how this can be.

@nrnhines From ef4da5d, I had assumed there would be a path to get from `hoc_PRintf` to `Printf` but it looks like it never calls `Printf`, which should be an alias to `nrnpy_pr`. If it never gets to `nrnpy_pr`, how does it show up inside a Jupyter notebook?

Reply:
My poor man's namespace is src/oc/redef.h ```# define	PRintf	hoc_PRintf``` and the former is in src/oc/fileio.h where ```PRintf``` uses ```NOT_PARALLEL_SUB(plprint(buf);)``` which sends to src/oc/plot.c and then to ```nrnpy_pr("%s", s);```

Reply:
So it does.

This pull request has now been merged, and the point isn't exactly how printf works, but is it worth pointing out (and saying that it's beyond the scope of the example) that this is potentially redirected to Python to do the actual printing?

Reply:
I would assume (all ?) functions that end up in the symbol table are redefined by suffixing `hoc_`.

My idea was to document the stack machine in action for a function call, because it makes use of all its parts (Datum stack, interpreter machine and frame stack). @salvadord thanks for the insight, I will specify that on my next iteration: I plan on finishing all topics on my list and have a 2nd run to update/improve them.

Reply:
> (all ?) functions ... redefined by suffixing hoc_
That is <supposed> to be the case.